### ISSUE TITLE
PlatboxDrv cannot be compiled on Linux

### ISSUE THREAD

**Issue by moldimolt** [IOActive/Platbox#8]
Hello. I cannot compile the kernetix driver on Linux:

```
[moldimolt@localhost-live driver]$ make
make -C /lib/modules/6.4.15-200.fc38.x86_64/build M=/home/moldimolt/Downloads/Platbox/PlatboxDrv/linux/driver clean
make -C /lib/modules/6.4.15-200.fc38.x86_64/build M=/home/moldimolt/Downloads/Platbox/PlatboxDrv/linux/driver modules
  CC [M]  /home/moldimolt/Downloads/Platbox/PlatboxDrv/linux/driver/kernetix.o
/home/moldimolt/Downloads/Platbox/PlatboxDrv/linux/driver/kernetix.c: In function ‘device_mmap’:
/home/moldimolt/Downloads/Platbox/PlatboxDrv/linux/driver/kernetix.c:64:23: error: assignment of read-only member ‘vm_flags’
   64 |         vma->vm_flags |= VM_IO;
      |                       ^~
/home/moldimolt/Downloads/Platbox/PlatboxDrv/linux/driver/kernetix.c:65:19: error: assignment of read-only member ‘vm_flags’
   65 |     vma->vm_flags |= (VM_DONTEXPAND | VM_DONTDUMP);
      |                   ^~
In file included from ./include/linux/kernel.h:30,
                 from ./arch/x86/include/asm/percpu.h:27,
                 from ./arch/x86/include/asm/nospec-branch.h:14,
                 from ./arch/x86/include/asm/paravirt_types.h:27,
                 from ./arch/x86/include/asm/ptrace.h:97,
                 from ./arch/x86/include/asm/math_emu.h:5,
                 from ./arch/x86/include/asm/processor.h:13,
                 from ./arch/x86/include/asm/timex.h:5,
                 from ./include/linux/timex.h:67,
                 from ./include/linux/time32.h:13,
                 from ./include/linux/time.h:60,
                 from ./include/linux/stat.h:19,
                 from ./include/linux/module.h:13,
                 from /home/moldimolt/Downloads/Platbox/PlatboxDrv/linux/driver/kernetix.c:6:
/home/moldimolt/Downloads/Platbox/PlatboxDrv/linux/driver/kernetix.c: In function ‘kernetix_ioctl’:
./include/linux/kern_levels.h:5:25: warning: format ‘%x’ expects argument of type ‘unsigned int’, but argument 2 has type ‘long unsigned int’ [-Wformat=]
    5 | #define KERN_SOH        "\001"          /* ASCII Start Of Header */
      |                         ^~~~~~
./include/linux/printk.h:427:25: note: in definition of macro ‘printk_index_wrap’
  427 |                 _p_func(_fmt, ##__VA_ARGS__);                           \
      |                         ^~~~
/home/moldimolt/Downloads/Platbox/PlatboxDrv/linux/driver/kernetix.c:90:9: note: in expansion of macro ‘printk’
   90 |         printk(KERN_ALERT "The IOCTL for ABW %#08x was called :)\n", KERNETIX_ABW);
      |         ^~~~~~
./include/linux/kern_levels.h:9:25: note: in expansion of macro ‘KERN_SOH’
    9 | #define KERN_ALERT      KERN_SOH "1"    /* action must be taken immediately */
      |                         ^~~~~~~~
/home/moldimolt/Downloads/Platbox/PlatboxDrv/linux/driver/kernetix.c:90:16: note: in expansion of macro ‘KERN_ALERT’
   90 |         printk(KERN_ALERT "The IOCTL for ABW %#08x was called :)\n", KERNETIX_ABW);
      |                ^~~~~~~~~~
./include/linux/kern_levels.h:5:25: warning: format ‘%x’ expects argument of type ‘unsigned int’, but argument 2 has type ‘long unsigned int’ [-Wformat=]
    5 | #define KERN_SOH        "\001"          /* ASCII Start Of Header */
      |                         ^~~~~~
./include/linux/printk.h:427:25: note: in definition of macro ‘printk_index_wrap’
  427 |                 _p_func(_fmt, ##__VA_ARGS__);                           \
      |                         ^~~~
/home/moldimolt/Downloads/Platbox/PlatboxDrv/linux/driver/kernetix.c:120:9: note: in expansion of macro ‘printk’
  120 |         printk(KERN_ALERT "The IOCTL for ABR %#08x was called :)\n", KERNETIX_ABR);
      |         ^~~~~~
./include/linux/kern_levels.h:9:25: note: in expansion of macro ‘KERN_SOH’
    9 | #define KERN_ALERT      KERN_SOH "1"    /* action must be taken immediately */
      |                         ^~~~~~~~
/home/moldimolt/Downloads/Platbox/PlatboxDrv/linux/driver/kernetix.c:120:16: note: in expansion of macro ‘KERN_ALERT’
  120 |         printk(KERN_ALERT "The IOCTL for ABR %#08x was called :)\n", KERNETIX_ABR);
      |                ^~~~~~~~~~
./include/linux/kern_levels.h:5:25: warning: format ‘%lx’ expects argument of type ‘long unsigned int’, but argument 2 has type ‘void *’ [-Wformat=]
    5 | #define KERN_SOH        "\001"          /* ASCII Start Of Header */
      |                         ^~~~~~
./include/linux/printk.h:427:25: note: in definition of macro ‘printk_index_wrap’
  427 |                 _p_func(_fmt, ##__VA_ARGS__);                           \
      |                         ^~~~
/home/moldimolt/Downloads/Platbox/PlatboxDrv/linux/driver/kernetix.c:126:9: note: in expansion of macro ‘printk’
  126 |         printk(KERN_ALERT "Value of %016lx --> %016lx\n",
      |         ^~~~~~
./include/linux/kern_levels.h:9:25: note: in expansion of macro ‘KERN_SOH’
    9 | #define KERN_ALERT      KERN_SOH "1"    /* action must be taken immediately */
      |                         ^~~~~~~~
/home/moldimolt/Downloads/Platbox/PlatboxDrv/linux/driver/kernetix.c:126:16: note: in expansion of macro ‘KERN_ALERT’
  126 |         printk(KERN_ALERT "Value of %016lx --> %016lx\n",
      |                ^~~~~~~~~~
./include/linux/kern_levels.h:5:25: warning: format ‘%lx’ expects argument of type ‘long unsigned int’, but argument 3 has type ‘void *’ [-Wformat=]
    5 | #define KERN_SOH        "\001"          /* ASCII Start Of Header */
      |                         ^~~~~~
./include/linux/printk.h:427:25: note: in definition of macro ‘printk_index_wrap’
  427 |                 _p_func(_fmt, ##__VA_ARGS__);                           \
      |                         ^~~~
/home/moldimolt/Downloads/Platbox/PlatboxDrv/linux/driver/kernetix.c:126:9: note: in expansion of macro ‘printk’
  126 |         printk(KERN_ALERT "Value of %016lx --> %016lx\n",
      |         ^~~~~~
./include/linux/kern_levels.h:9:25: note: in expansion of macro ‘KERN_SOH’
    9 | #define KERN_ALERT      KERN_SOH "1"    /* action must be taken immediately */
      |                         ^~~~~~~~
/home/moldimolt/Downloads/Platbox/PlatboxDrv/linux/driver/kernetix.c:126:16: note: in expansion of macro ‘KERN_ALERT’
  126 |         printk(KERN_ALERT "Value of %016lx --> %016lx\n",
      |                ^~~~~~~~~~
./include/linux/kern_levels.h:5:25: warning: format ‘%x’ expects argument of type ‘unsigned int’, but argument 2 has type ‘long unsigned int’ [-Wformat=]
    5 | #define KERN_SOH        "\001"          /* ASCII Start Of Header */
      |                         ^~~~~~
./include/linux/printk.h:427:25: note: in definition of macro ‘printk_index_wrap’
  427 |                 _p_func(_fmt, ##__VA_ARGS__);                           \
      |                         ^~~~
/home/moldimolt/Downloads/Platbox/PlatboxDrv/linux/driver/kernetix.c:136:9: note: in expansion of macro ‘printk’
  136 |         printk(KERN_ALERT "The IOCTL for CR3 %#08x was called :)\n", KERNETIX_CR3);
      |         ^~~~~~
./include/linux/kern_levels.h:9:25: note: in expansion of macro ‘KERN_SOH’
    9 | #define KERN_ALERT      KERN_SOH "1"    /* action must be taken immediately */
      |                         ^~~~~~~~
/home/moldimolt/Downloads/Platbox/PlatboxDrv/linux/driver/kernetix.c:136:16: note: in expansion of macro ‘KERN_ALERT’
  136 |         printk(KERN_ALERT "The IOCTL for CR3 %#08x was called :)\n", KERNETIX_CR3);
      |                ^~~~~~~~~~
./include/linux/kern_levels.h:5:25: warning: format ‘%p’ expects argument of type ‘void *’, but argument 2 has type ‘long unsigned int’ [-Wformat=]
    5 | #define KERN_SOH        "\001"          /* ASCII Start Of Header */
      |                         ^~~~~~
./include/linux/printk.h:427:25: note: in definition of macro ‘printk_index_wrap’
  427 |                 _p_func(_fmt, ##__VA_ARGS__);                           \
      |                         ^~~~
/home/moldimolt/Downloads/Platbox/PlatboxDrv/linux/driver/kernetix.c:145:9: note: in expansion of macro ‘printk’
  145 |         printk(KERN_ALERT "Value of CR3 (Physical Address) --> %p\n", _cr3);
      |         ^~~~~~
./include/linux/kern_levels.h:9:25: note: in expansion of macro ‘KERN_SOH’
    9 | #define KERN_ALERT      KERN_SOH "1"    /* action must be taken immediately */
      |                         ^~~~~~~~
/home/moldimolt/Downloads/Platbox/PlatboxDrv/linux/driver/kernetix.c:145:16: note: in expansion of macro ‘KERN_ALERT’
  145 |         printk(KERN_ALERT "Value of CR3 (Physical Address) --> %p\n", _cr3);
      |                ^~~~~~~~~~
In file included from ./include/linux/uaccess.h:11,
                 from ./include/linux/sched/task.h:11,
                 from ./include/linux/sched/signal.h:9,
                 from ./include/linux/rcuwait.h:6,
                 from ./include/linux/percpu-rwsem.h:7,
                 from ./include/linux/fs.h:33,
                 from /home/moldimolt/Downloads/Platbox/PlatboxDrv/linux/driver/kernetix.c:9:
./arch/x86/include/asm/uaccess.h:175:34: warning: initialization of ‘long unsigned int *’ from ‘long unsigned int’ makes pointer from integer without a cast [-Wint-conversion]
  175 |         __typeof__(*(ptr)) __x = (x); /* eval x once */                 \
      |                                  ^
./arch/x86/include/asm/uaccess.h:208:44: note: in expansion of macro ‘do_put_user_call’
  208 | #define put_user(x, ptr) ({ might_fault(); do_put_user_call(put_user,x,ptr); })
      |                                            ^~~~~~~~~~~~~~~~
/home/moldimolt/Downloads/Platbox/PlatboxDrv/linux/driver/kernetix.c:146:13: note: in expansion of macro ‘put_user’
  146 |             put_user( _cr3, (unsigned long **)p );
      |             ^~~~~~~~
/home/moldimolt/Downloads/Platbox/PlatboxDrv/linux/driver/kernetix.c:174:9: warning: ISO C90 forbids mixed declarations and code [-Wdeclaration-after-statement]
  174 |         __u64 val = 0;
      |         ^~~~~
./arch/x86/include/asm/uaccess.h:175:34: warning: initialization of ‘long unsigned int *’ from ‘__u64’ {aka ‘long long unsigned int’} makes pointer from integer without a cast [-Wint-conversion]
  175 |         __typeof__(*(ptr)) __x = (x); /* eval x once */                 \
      |                                  ^
./arch/x86/include/asm/uaccess.h:208:44: note: in expansion of macro ‘do_put_user_call’
  208 | #define put_user(x, ptr) ({ might_fault(); do_put_user_call(put_user,x,ptr); })
      |                                            ^~~~~~~~~~~~~~~~
/home/moldimolt/Downloads/Platbox/PlatboxDrv/linux/driver/kernetix.c:176:9: note: in expansion of macro ‘put_user’
  176 |         put_user( val, (unsigned long **)msr_call->result );
      |         ^~~~~~~~
/home/moldimolt/Downloads/Platbox/PlatboxDrv/linux/driver/kernetix.c:210:9: warning: ISO C90 forbids mixed declarations and code [-Wdeclaration-after-statement]
  210 |         UINT32 bus      = pci_call->bus;
      |         ^~~~~~
./arch/x86/include/asm/uaccess.h:175:34: warning: initialization of ‘long unsigned int *’ from ‘BYTE’ {aka ‘unsigned char’} makes pointer from integer without a cast [-Wint-conversion]
  175 |         __typeof__(*(ptr)) __x = (x); /* eval x once */                 \
      |                                  ^
./arch/x86/include/asm/uaccess.h:208:44: note: in expansion of macro ‘do_put_user_call’
  208 | #define put_user(x, ptr) ({ might_fault(); do_put_user_call(put_user,x,ptr); })
      |                                            ^~~~~~~~~~~~~~~~
/home/moldimolt/Downloads/Platbox/PlatboxDrv/linux/driver/kernetix.c:219:9: note: in expansion of macro ‘put_user’
  219 |         put_user( val, (unsigned long **)pci_call->result );
      |         ^~~~~~~~
/home/moldimolt/Downloads/Platbox/PlatboxDrv/linux/driver/kernetix.c:234:9: warning: ISO C90 forbids mixed declarations and code [-Wdeclaration-after-statement]
  234 |         UINT32 bus      = pci_call->bus;
      |         ^~~~~~
./arch/x86/include/asm/uaccess.h:175:34: warning: initialization of ‘long unsigned int *’ from ‘WORD’ {aka ‘short unsigned int’} makes pointer from integer without a cast [-Wint-conversion]
  175 |         __typeof__(*(ptr)) __x = (x); /* eval x once */                 \
      |                                  ^
./arch/x86/include/asm/uaccess.h:208:44: note: in expansion of macro ‘do_put_user_call’
  208 | #define put_user(x, ptr) ({ might_fault(); do_put_user_call(put_user,x,ptr); })
      |                                            ^~~~~~~~~~~~~~~~
/home/moldimolt/Downloads/Platbox/PlatboxDrv/linux/driver/kernetix.c:243:9: note: in expansion of macro ‘put_user’
  243 |         put_user( val, (unsigned long **)pci_call->result );
      |         ^~~~~~~~
/home/moldimolt/Downloads/Platbox/PlatboxDrv/linux/driver/kernetix.c:258:9: warning: ISO C90 forbids mixed declarations and code [-Wdeclaration-after-statement]
  258 |         UINT32 bus      = pci_call->bus;
      |         ^~~~~~
./arch/x86/include/asm/uaccess.h:175:34: warning: initialization of ‘long unsigned int *’ from ‘DWORD’ {aka ‘unsigned int’} makes pointer from integer without a cast [-Wint-conversion]
  175 |         __typeof__(*(ptr)) __x = (x); /* eval x once */                 \
      |                                  ^
./arch/x86/include/asm/uaccess.h:208:44: note: in expansion of macro ‘do_put_user_call’
  208 | #define put_user(x, ptr) ({ might_fault(); do_put_user_call(put_user,x,ptr); })
      |                                            ^~~~~~~~~~~~~~~~
/home/moldimolt/Downloads/Platbox/PlatboxDrv/linux/driver/kernetix.c:267:9: note: in expansion of macro ‘put_user’
  267 |         put_user( val, (unsigned long **)pci_call->result );
      |         ^~~~~~~~
/home/moldimolt/Downloads/Platbox/PlatboxDrv/linux/driver/kernetix.c:282:9: warning: ISO C90 forbids mixed declarations and code [-Wdeclaration-after-statement]
  282 |         UINT32 bus      = pci_call->bus;
      |         ^~~~~~
/home/moldimolt/Downloads/Platbox/PlatboxDrv/linux/driver/kernetix.c:303:9: warning: ISO C90 forbids mixed declarations and code [-Wdeclaration-after-statement]
  303 |         UINT32 bus      = pci_call->bus;
      |         ^~~~~~
/home/moldimolt/Downloads/Platbox/PlatboxDrv/linux/driver/kernetix.c:324:9: warning: ISO C90 forbids mixed declarations and code [-Wdeclaration-after-statement]
  324 |         UINT32 bus      = pci_call->bus;
      |         ^~~~~~
/home/moldimolt/Downloads/Platbox/PlatboxDrv/linux/driver/kernetix.c:345:9: warning: ISO C90 forbids mixed declarations and code [-Wdeclaration-after-statement]
  345 |         UINT32 bus      = pci_call->bus;
      |         ^~~~~~
/home/moldimolt/Downloads/Platbox/PlatboxDrv/linux/driver/kernetix.c:370:9: warning: ISO C90 forbids mixed declarations and code [-Wdeclaration-after-statement]
  370 |         UINT32 bus      = pci_call->bus;
      |         ^~~~~~
./arch/x86/include/asm/uaccess.h:175:34: warning: initialization of ‘long unsigned int *’ from ‘UINT64’ {aka ‘long long unsigned int’} makes pointer from integer without a cast [-Wint-conversion]
  175 |         __typeof__(*(ptr)) __x = (x); /* eval x once */                 \
      |                                  ^
./arch/x86/include/asm/uaccess.h:208:44: note: in expansion of macro ‘do_put_user_call’
  208 | #define put_user(x, ptr) ({ might_fault(); do_put_user_call(put_user,x,ptr); })
      |                                            ^~~~~~~~~~~~~~~~
/home/moldimolt/Downloads/Platbox/PlatboxDrv/linux/driver/kernetix.c:379:9: note: in expansion of macro ‘put_user’
  379 |         put_user( bar_size, (unsigned long **)pci_call->result );
      |         ^~~~~~~~
In file included from ./include/linux/linkage.h:7,
                 from ./arch/x86/include/asm/cache.h:5,
                 from ./include/linux/cache.h:6,
                 from ./include/linux/time.h:5:
/home/moldimolt/Downloads/Platbox/PlatboxDrv/linux/driver/kernetix.c: In function ‘driver_initialize’:
./include/linux/export.h:27:22: error: passing argument 1 of ‘class_create’ from incompatible pointer type [-Werror=incompatible-pointer-types]
   27 | #define THIS_MODULE (&__this_module)
      |                     ~^~~~~~~~~~~~~~~
      |                      |
      |                      struct module *
/home/moldimolt/Downloads/Platbox/PlatboxDrv/linux/driver/kernetix.c:494:31: note: in expansion of macro ‘THIS_MODULE’
  494 |   device_class = class_create(THIS_MODULE, KERNETIX_DEVICE_NAME);
      |                               ^~~~~~~~~~~
In file included from ./include/linux/device.h:31,
                 from ./include/linux/cdev.h:8,
                 from /home/moldimolt/Downloads/Platbox/PlatboxDrv/linux/driver/kernetix.c:13:
./include/linux/device/class.h:230:54: note: expected ‘const char *’ but argument is of type ‘struct module *’
  230 | struct class * __must_check class_create(const char *name);
      |                                          ~~~~~~~~~~~~^~~~
/home/moldimolt/Downloads/Platbox/PlatboxDrv/linux/driver/kernetix.c:494:18: error: too many arguments to function ‘class_create’
  494 |   device_class = class_create(THIS_MODULE, KERNETIX_DEVICE_NAME);
      |                  ^~~~~~~~~~~~
./include/linux/device/class.h:230:29: note: declared here
  230 | struct class * __must_check class_create(const char *name);
      |                             ^~~~~~~~~~~~
cc1: some warnings being treated as errors
make[2]: *** [scripts/Makefile.build:252: /home/moldimolt/Downloads/Platbox/PlatboxDrv/linux/driver/kernetix.o] Error 1
make[1]: *** [Makefile:2050: /home/moldimolt/Downloads/Platbox/PlatboxDrv/linux/driver] Error 2
make: *** [Makefile:25: kernetix] Error 2
```

Thank you.

**Comment by droogie** [comment]
Looks like Linux Kernel changes for the latest releases have changed build requirements, for example can't seem to directly modify vm_flags anymore and are expected to use these functions in mm.h

`
	/*
	 * Flags, see mm.h.
	 * To modify use vm_flags_{init|reset|set|clear|mod} functions.
	 */
	union {
		const vm_flags_t vm_flags;
		vm_flags_t __private __vm_flags;
	};
`

Just tried compiling on 6.5.5 and ran into the same issue as you. For now, you could boot to a live Linux environment on the platform you're targeting running an older kernel and it should build for you out of the box

**Comment by droogie** [comment]
Just fixed this in the latest commit (909a7f63e66068f10d7481a160c8bbdcf38f05e1), turned out to be from two different changes in the Linux Kernel, one in 6.3 and one in 6.4
